fix a few things

FireBird702 opened the pull request.

Commit:
fix keep:Overwrite() not counting towards ._activeSaveJobs

FireBird702 commented:
Also I've noticed that `:Overwrite()` adds `.ActiveSession` to `.MetaData` and doesn't removes it. Idk if it's a bug or not

noahrepublic commented:
> Also I've noticed that `:Overwrite()` adds `.ActiveSession` to `.MetaData` and doesn't removes it. Idk if it's a bug or not

I think my original thought was that :Overwrite should also steal the session, but I don't think this is good. So yes it is a bug.

FireBird702 commented:
> I think my original thought was that :Overwrite should also steal the session, but I don't think this is good. So yes it is a bug.

Wouldn't the changes be overwritten again if `:Overwrite()` is called on a keep that is session locked on the other server?

noahrepublic commented:
> > I think my original thought was that :Overwrite should also steal the session, but I don't think this is good. So yes it is a bug.
> 
> Wouldn't the changes be overwritten again if `:Overwrite()` is called on a keep that is session locked on the other server?

Seems there is a bigger problem then we thought then. I can't currently look at specifically, but I can discuss possible solutions

FireBird702 commented:
> > Wouldn't the changes be overwritten again if `:Overwrite()` is called on a keep that is session locked on the other server?
> 
> Seems there is a bigger problem then we thought then. I can't currently look at specifically, but I can discuss possible solutions

Maybe a way to tell the other server to release the session without saving when `:Overwrite()` is called.

FireBird702 commented:
`:Overwrite()` will set `.Overwritten` in `.MetaData` to `true`, already locked session during auto save will do a check and do a `:Release()` removing session lock but not saving data. `:LoadKeep()` will set `.Overwritten` to `nil`. Still unsure what if `:Overwrite()` gets called, other session still waits for auto save check, and `:LoadKeep()` is called somewhere. `:LoadKeep()` releases session from the other server if force loaded?

noahrepublic commented:
> `:Overwrite()` will set `.Overwritten` in `.MetaData` to `true`, already locked session during auto save will do a check and do a `:Release()` removing session lock but not saving data. `:LoadKeep()` will set `.Overwritten` to `nil`. Still unsure what if `:Overwrite()` gets called, other session still waits for auto save check, and `:LoadKeep()` is called somewhere. `:LoadKeep()` releases session from the other server if force loaded?

This seems good but requires a lot of change in conditions, we should take a look at what profileservice does

Commit:
fix ForceLoad, keep.Releasing, keepStore.validate() and incorrect JobID type, change what :Save() and :Overwrite() returns

FireBird702 commented:
Changes:
- fixed `keepStore.validate()` using incorrect data
- fixed `keep.Relasing` not firing
- Removed `Ignore`, added `ForceLoad` to `unreleasedHandler`. Maybe add true `Ignore`? 🤔
- fixed `ForceLoad` (previously `Ignore`) not working correctly
- `:Save()`, `:Overwrite()` now returns `keep` instead of `DataStoreKeyInfo` (`keep:GetKeyInfo()` exists), `:Release()` already does that. I can revert that if there is a specific use case
- fixed incorrect JobID type

FireBird702 commented:
> This seems good but requires a lot of change in conditions, we should take a look at what profileservice does

ProfileService has the same problem from what I see.

Copied from discord:
Maybe we can do something similar to ForceLoad in my latest commit in pr. Because `:Overwrite()` uses `:UpdateAsync()` we can check if session is locked somewhere and tell to release the session without saving

Commit:
:Overwrite() should release locked session, fix for ForceLoad still setting data

FireBird702 commented:
Now you can see what I meant

Commit:
Added a way to control `Keep:Overwrite()` behavior, added .Overwritten signal, fixed .LastUpdate not applying, some other fixes

FireBird702 commented:
Should be ready for review

Commit:
fix error in keep in mock store

noahrepublic commented:
Curious what part of "Forceload" (or Ignore) was not working

I will review in about 20 minutes

noahrepublic commented on src/Keep.lua:
```diff
@@ -286,27 +296,48 @@ local function isKeepLocked(metaData: MetaData)
 				local keepStore = keep._keep_store
 
-				local valid, err = keepStore.validate(newestData.Data)
+				local valid, err = keepStore.validate(keep.Data) -- validate data before saving
```
Unsure what i was doing here...

noahrepublic commented on src/Keep.lua:
```diff
@@ -326,10 +357,7 @@ local function transformUpdate(keep: Keep, newestData: KeepStruct, release: bool
-				ID = 0,
-				Updates = {},
-			} -- the final global updates to save
+			local finalGlobals = { ID = 0, Updates = {} } -- the final global updates to save
```
These spanned out tables may always exist. I have stylua to auto format tables for easier to read

noahrepublic commented on src/Keep.lua:
```diff
@@ -419,79 +447,133 @@ local function transformUpdate(keep: Keep, newestData: KeepStruct, release: bool
+
+	if self._released then
+		return
+	end
```
```
        if releaseCache[self:Identify()] then -- already releasing
		return releaseCache[self:Identify()]
	end

	if self._released then
		return
	end
```

Isn't this redundant since these same checks work made in ```:Release``

FireBird702 replied:
I've made it just in case for session release during saving. `:Release()` checks it before anything but overwriting and force loading checks that in `:_release()`

noahrepublic commented on src/init.lua:
```diff
@@ -480,71 +484,63 @@ end
-		and self._cachedKeepPromises[identifier].Status ~= Promise.Status.Cancelled
-	then
-		return self._cachedKeepPromises[identifier]
+	local id = `{self._store_info.Name}/{self._store_info.Scope or ""}{self._store_info.Scope and "/" or ""}{key}`
```
This seems messier imo 🤷‍♂️

FireBird702 replied:
It looks for me better because it takes less space to read it but 🤷‍♂️

noahrepublic commented on src/init.lua:
```diff
@@ -480,71 +484,63 @@ end
 
-			if loadMethod == "Cancel" then
+			if loadMethod == validLoadMethods.Cancel then
 				reject(nil) -- should this return an error object?
```
We should probably check if the session is past the deadlock time.

noahrepublic commented on src/init.lua:
```diff
@@ -480,71 +484,63 @@ end
-			if loadMethod == "Cancel" then
+			if loadMethod == validLoadMethods.Cancel then
 				reject(nil) -- should this return an error object?
 				return
```
When ``Cancel`` is passed as the handler, we should probably check if the session that owns it is past deadlock time. I would probably assume this is intended behavior if I read the docs

FireBird702 replied:
`canLoad()` already does that

noahrepublic commented on src/init.lua:
```diff
@@ -480,71 +484,63 @@ end
+		ForceLoad = "ForceLoad",
+		-- maybe add "Ignore"?
+		Cancel = "Cancel",
+	}
```
Maybe could globalize this or even a type if possible. Making it accessible may help users if docs aren't clear enough.

FireBird702 replied:
I think it could be helpful

noahrepublic commented on stylua.toml:
```diff
@@ -0,0 +1,10 @@
+collapse_simple_statement = "Never"
+
+[sort_requires]
+enabled = true
```
Was just thinking about this. I should make an action to auto-format things to fix any inconsistency

noahrepublic commented on src/Keep.lua:
```diff
@@ -758,47 +831,21 @@ function Keep:Release()
-	releaseCache[self:Identify()] = updater
-
-	return updater
+	return self:_release(updater)
```
Maybe I am missing something, but we lose the functionality of firing ``.Releasing`` here? I probably am saying this and you have already fixed it in another commit

Edit: I see ``:_release`` fires it, but seems it isn't under the same condition. Maybe the condition is useless and I am missing something?

FireBird702 replied:
I don't see any problems here but I'll check that later today

Commit:
`.LoadMethods` are now stored in DataKeep and accessible externally, more info when using mock store in `:LoadKeep()`, added `[DataKeep]` to log messages to better indicate they are from DataKeep

FireBird702 commented:
> Curious what part of "Forceload" (or Ignore) was not working

It was a weird mix of "Ignore" and "ForceLoad" where nothing was working properly

FireBird702 committed:
change to `:expect()`

noahrepublic reviewed (commented):
Seems good, now with the renaming of ``Ignore`` to ``Forceload`` we should add a proper ignore for next patch.

Just running the last tests before merging.

noahrepublic reviewed (changes requested):
I noticed your commits are two behind the main because of some doc changes. Worried if those will merge correctly. but aside from that ready to merge.

FireBird702 committed:
Merge branch 'noahrepublic:main' into main
